CRM/Contribute/BAO/Contribution.php

```diff
@@ -4519,10 +4550,38 @@ public static function isSingleLineItem($id) {
    * @param int $recur
    * @param CRM_Contribute_BAO_Contribution $contribution
    *
-   * @return array
+   * @return NULL|array
    */
   public static function completeOrder(&$input, &$ids, $objects, $transaction, $recur, $contribution) {
     $primaryContributionID = isset($contribution->id) ? $contribution->id : $objects['first_contribution']->id;
+    $paymentProcessorId = '';
```

Comment:
OK - so this is my big issue here - the completeOrder function should NOT be handling partial payments. It is the function to act when a contribution has been completed and the various things need to be transitioned.

The way we should be developing here is building out the Payment.create api - which calls this completeOrder function when it is the final payment - so ideally we want the form function

completeTransaction

to call Payment.create instead of contribution.Completetransaction - which will add payments & complete the transaction if required.

Currently Payment.create currently has some gaps :-( but I think we should be addresssing those. We definitely should not be squeezing Payment.create functionality into completetransaction / completeOrder

Reply:
As mentioned above once, we can make the existing fn ```CRM_Contribute_BAO_Contribution::recordAdditonalPayment(..)``` more robust to handle partial and even refund payment properly then we can call this fn in Payment.Create API, in this way we can differentiate it from completetransaction

Reply:
Hmm, concur with the improvement of the payment create api to include more flexibility in accepting the params 👍  

As per the changes in this PR - looking at the screenshot in the PR description and some following comments - the front end form available to the user displays a non-editable balance amount which on submission do completes the payment for the contribution, hence the logic was kept same as we did the implementation for completing `pending` payments.

But, as the backend form offers a set of periodic payments until a contribution is completed (eg $100 is completed in $20, $30 and then $50) - agree this should also be added for front end forms too - which was also asked by our client after this PR and we implemented for them but thought to create a new PR for that and not mess this up as it had already completed some testing by Richard and others.

I can add the improvements related to create payment api, etc by the end of this week 👍 

Thanks Eileen and Monish for the comments.

Reply:
So my feeling is we should focus on getting our basics right first. Ie. we want to get the payment api working well enough that we can replace all instances where a payment is received with Payment.create (this would include the many places where we currently call completetransaction).

Fortunately we have built up pretty substantial test 'assets' in most of the places that receive payments.

I would like to propose that we work on maturing the Payment.create api by switching the AdditionalPayment form over to using it as a first step. I did a [commit  here](https://github.com/eileenmcnaughton/civicrm-core/commit/431a5909dc800459d2de3a9626d3b0d0a51e340b) - the current failure reason is some sillyness in the Payment create api itself (it's not updating status when you update a partially paid) but I think we can add a test directly against the Payment.create api that captures that scenario & fixes it.



I think really cleaning up Payment.create & working together on switching forms to it will set us up a lot better in the future  - both while we are working with quick form & for when we finally escape it.

Reply:
Updated the PR to use the payment create API for partial payments on front and backend forms. I've also included the above commit from @eileenmcnaughton. Till now, I don't see any issues, but changing this to wip until the following point is completely checked.

- Test payment create api for any gaps and handle them. Eg. it does not send email receipt from front-end forms after the thankyou page is loaded.
- Handle the assignment of `Civi::settings()->get('contribution_invoice_settings');`
- Add more unit test to assert the API results.